### ISSUE TITLE
compiler fails to reject a program

### ISSUE THREAD

**Issue by hemmecke** [fricas/fricas#164]
The following program is missing `==`  at position (*), but still compiles fine.
I only detected that problem, because, when I wanted to use `uNCoefficient`, the interpreter claimed that that function is not defined in `QEtaOneOverPi(Integer)`.

The only hint I can see is:

```
   Semantic Errors: 
      [1]  C is not a known type
```
I wonder by which reasoning the compiler is able to accept that program.

Here is the program
```
PP ==> PositiveInteger
ZZ ==> Integer
MZZ ==> Matrix ZZ -- consider only 2x2 matricies
)abbrev package Q1PI QEtaOneOverPi
QEtaOneOverPi(C: IntegralDomain): Exports == Implementation where
  Exports ==> with
    uNCoefficient: (PP, MZZ, C) -> C
    if C has Field then
        moebiusTransform: (MZZ, C) -> C
  Implementation ==> add
    uNCoefficient(nn: PP, gamma: MZZ, tau: C): C -----------------------------(*)
        c: ZZ := gamma(2,1)
        d: ZZ := gamma(2,2)
        determinant(gamma) - nn*(c*tau+d)^2
    if C has Field then
        moebiusTransform(gamma: MZZ, x: C): C ==
            (gamma(1,1)*x+gamma(1,2)::C) / (gamma(2,1)*x+gamma(2,2)::C)
```
If you remove the last `iF` and all lines below, then the compiler says
```
   QEtaOneOverPi has 1 precompilation error:
    Parsing error: Apparent indentation error following add 

 
   >> Apparent user error:
   precompilation failed
```

**Comment by hebisch** [comment]
On Mon, Apr 08, 2024 at 06:03:06AM -0700, Ralf Hemmecke wrote:
> The following program is missing `==`  at position (*), but still compiles fine.
> I only detected that problem, because, when I wanted to use `uNCoefficient`, the interpreter claimed that that function is not defined in `QEtaOneOverPi(Integer)`.
> 
> The only hint I can see is:
> 
> ```
>    Semantic Errors: 
>       [1]  C is not a known type
> ```
> I wonder by which reasoning the compiler is able to accept that program.
> 
> Here is the program
> ```
> PP ==> PositiveInteger
> ZZ ==> Integer
> MZZ ==> Matrix ZZ -- consider only 2x2 matricies
> )abbrev package Q1PI QEtaOneOverPi
> QEtaOneOverPi(C: IntegralDomain): Exports == Implementation where
>   Exports ==> with
>     uNCoefficient: (PP, MZZ, C) -> C
>     if C has Field then
>         moebiusTransform: (MZZ, C) -> C
>   Implementation ==> add
>     uNCoefficient(nn: PP, gamma: MZZ, tau: C): C -----------------------------(*)
>         c: ZZ := gamma(2,1)
>         d: ZZ := gamma(2,2)
>         determinant(gamma) - nn*(c*tau+d)^2
>     if C has Field then
>         moebiusTransform(gamma: MZZ, x: C): C ==
>             (gamma(1,1)*x+gamma(1,2)::C) / (gamma(2,1)*x+gamma(2,2)::C)
> ```
> If you remove the last `iF` and all lines below, then the compiler says
> ```
>    QEtaOneOverPi has 1 precompilation error:
>     Parsing error: Apparent indentation error following add 
> 
>  
>    >> Apparent user error:
>    precompilation failed
> ```
 
The
```
   >> Apparent user error:
   precompilation failed
```
looks like a compiler bug, AFAICS the code is syntactically correct.
Correct parse of your program is as follows: you have _declaration_
of function 'uNCoefficient' with return type which is C applied
to a pile.  The

```
  Semantic Errors:
      [1]  C is not a known type
```
message points to the problem: compiler does not know about one
argument type 'C'.  Maybe this error should cause compile failure.
Currently during algebra compilation this error appears 7 times,
we should be able to eliminate all such errors.  Also, currently
overloading machinery distinguishes meaning as a variable/parameter
and meaning as a type and apparenatly allows both.  Arguably compiler
could treat use as variable and use as a type as a conflict causing
compile failure.

-- 
                              Waldek Hebisch

**Comment by hemmecke** [comment]
Yes, syntacticallly it is C applied to the block and the whole thing is a declaration.
However, with the information at hand, C is an integral domain and not a functor that can be applied to something.

I do not quite understand the stuff with "disguises meaning as a variable/parameter". I do not see how the compiler can generate reasonable code for this "declaration", because the compiler should be unable to figure out what "C(some_block)" means here.

It should definitely abort the compilation process at that place.

**Comment by hebisch** [comment]
On Mon, Apr 08, 2024 at 12:36:53PM -0700, Ralf Hemmecke wrote:
> Yes, syntacticallly it is C applied to the block and the whole thing is a declaration.
> However, with the information at hand, C is an integral domain and not a functor that can be applied to something.
> 
> I do not quite understand the stuff with "disguises meaning as a
                                            ^^^^^^^^^
distinguishes
> variable/parameter". I do not see how the compiler can generate reasonable code for this "declaration", because the compiler should be unable to figure out what "C(some_block)" means here.

Well, this "declaration" is unsed, so no code to generate.  AFAIU
your domain worked fine, except for fact that 'uNCoefficient' was
not defined.

-- 
                              Waldek Hebisch

**Comment by hemmecke** [comment]
OK. "distinguish", but I still don't get it why in `C(some_block)` there is "variable/parameter" involved. And anyway, I do not see, how it can make any sense, since the only C that is know in this context is `C: IntegralDomain`.

That there is dead-code-removal, is clear, but before that I would have thought that the compiler has to make some sense out of the code characters. Isn't there some type checking involved in that case? `C(some_block)` must be a type (a domain preferably), but how could the compiler confirm this?
